sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -464,6 +465,24 @@ object OptimizeIn extends Rule[LogicalPlan] {
 }
 
 /**
+  * Convert an expression into its conjunctive normal form (CNF), i.e. AND of ORs.
+  * For example, a && b || c is normalized to (a || c) && (b || c) by this method.
+  *
+  * Refer to https://en.wikipedia.org/wiki/Conjunctive_normal_form for more information
+  */
+object CNFNormalization extends Rule[LogicalPlan] {
```

Comment:
How about using the following heuristic solution to prevent exponential explosion:

1.  Add a simple `size` method to `TreeNode`, which returns the size (total number of nodes) of a tree:

    ```scala
    def size: Int = 1 + children.map(_.size).sum
    ```

1.  Gives up CNF conversion once the result predicate exceeds a predefined threshold.

    For example, we can stop if the size of the converted predicate is 10 times larger than the original one.